Add Initial Java Support for GDS to KvikIO

aslobodaNV opened the pull request:
This PR is intended to add initial support for Java binding to GDS as part of the KvikIO library. In this PR are the minimal set of bindings required to support synchronous read and write IO operations via GDS as well as a single example to demonstrate how the bindings can be used alongside other CUDA libraries, such as JCuda. Full support for the GDS CuFile API, including batch and asynchronous IO, has not yet been implemented and more sophisticated error/exception handling is not yet in place. There is a README located within kvikio/java detailing how this new functionality can be compiled and built locally, along with detailed instructions on how to run the included usage example.

madsbk commented:
/ok to test

wence- commented on java/README.md:
```diff
@@ -0,0 +1,48 @@
+Java Bindings
+
+Summary
+These Java KvikIO bindings for GDS currently support only synchronous read and write IO operations using the underlying CuFile API. Support for batch IO and asynchronous operations are not yet supported.
+
+Dependencies
+The Java KvikIO bindings have been developed to work on Linux based systems and require CUDA to be installed and for GDS to be properly enabled. Instructions for how to install and enable GDS can be found on NVIDIA's website. To compile the shared library it is also necessary to have a JDK installed. To run the included example, it is also necessary to install JCuda as it is used to handle memory ...
+https://repo1.maven.org/maven2/org/jcuda/jcuda/12.0.0/jcuda-12.0.0.jar
+https://repo1.maven.org/maven2/org/jcuda/jcuda-natives/12.0.0/jcuda-natives-12.0.0.jar
+
+Compilation
+To recompile the .so file for your local system run the following command. Note: Update the command to reflect the directory where you have installed CUDA and your JDK.
+
```
Let's use markdown formatting here, so that this will look a bit better when rendered.

```suggestion
# Java Bindings

## Summary
These Java KvikIO bindings for GDS currently support only synchronous read and write IO operations using the underlying CuFile API. Support for batch IO and asynchronous operations are not yet supported.

## Dependencies
The Java KvikIO bindings have been developed to work on Linux based systems and require CUDA to be installed and for GDS to be properly enabled. Instructions for how to install and enable GDS can be found on NVIDIA's website. To compile the shared library it is also necessary to have a JDK installed. To run the included example, it is also necessary to install JCuda as it is used to handle memory allocations and the transfer of data between host and GPU memory. JCuda jar files supporting CUDA 12.x can be found here:
https://repo1.maven.org/maven2/org/jcuda/jcuda/12.0.0/jcuda-12.0.0.jar
https://repo1.maven.org/maven2/org/jcuda/jcuda-natives/12.0.0/jcuda-natives-12.0.0.jar

## Compilation
To recompile the .so file for your local system run the following command. Note: Update the command to reflect the directory where you have installed CUDA and your JDK.

```

(See https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax for further syntax).

wence- commented on java/README.md:
```diff
@@ -0,0 +1,48 @@
+
+Dependencies
+The Java KvikIO bindings have been developed to work on Linux based systems and require CUDA to be installed and for GDS to be properly enabled. Instructions for how to install and enable GDS can be found on NVIDIA's website. To compile the shared library it is also necessary to have a JDK installed. To run the included example, it is also necessary to install JCuda as it is used to handle memory ...
+https://repo1.maven.org/maven2/org/jcuda/jcuda/12.0.0/jcuda-12.0.0.jar
+https://repo1.maven.org/maven2/org/jcuda/jcuda-natives/12.0.0/jcuda-natives-12.0.0.jar
```
*suggestion*: Prefer not to link to explicit files since this can easily go out of date without us remembering to update things. For preference, can we link to an appropriate installation document for JCuda?

aslobodaNV replied:
Thanks for the suggestion, I can link to the installation instructions as well but found that the installation instructions themselves are a bit unclear/potentially out of date. I may keep these links but just add a note to confirm they are still up to date and reference the installation instructions.

wence- commented on java/src/main/native/src/CuFileJni.cpp:
I am a bit confused by these bindings. They look to be bindings to the cufile API, and not the kvikio API. Is that deliberate? It seems like if so one will have to replicate a lot of the implementation that already exists in kvikio here.

aslobodaNV replied:
You are correct, these are directly bound to the underlying cufile API. In the initial discussions with the rest of the GDS team there was some concern about potential performance overhead of Java bindings -> C++ bindings -> C API. I have not investigated the overhead yet, but definitely see the benefit you are suggesting to sharing one underlying set of bindings. If you don't mind, I will add an issue to my backlog to investigate the potential performance differences and, if they are minimal, update these to point to the C++ bindings at a later date.

madsbk replied:
Sounds good, let's continue with raw java bindings for now. 

Side note, I would be very surprised if the basic C++ API introduce any significant overhead. E.g. the overhead of calling [`read()`](https://github.com/rapidsai/kvikio/blob/branch-24.08/cpp/include/kvikio/file_handle.hpp#L301) vs. calling `cuFileRead` is tiny.

madsbk commented:
/ok to test

bdice reviewed (changes requested):
A few high level comments before digging too deep into this code:

- I have a weak preference to wrap the C++ bindings rather than re-implement kvikio in Java. This will probably be easier than reimplementing the code as new features are introduced to kvikio, and gives more confidence in testing (bindings are probably less bug-prone than a full reimplementation).
- This will need to pass style checks. See the [cuDF guide on Code Formatting](https://github.com/rapidsai/cudf/blob/branch-24.08/CONTRIBUTING.md#code-formatting). The same guidance applies to using `pre-commit` with kvikio development.
- This will need a build system, probably Maven. Long commands in the README that explain how to supply all the dependencies are not a good solution if those dependencies can be fetched and built automatically. See how cuDF's Java bindings use Maven, for instance. https://github.com/rapidsai/cudf/tree/branch-24.08/java
  - The end result should be that commands like `mvn clean install` work properly.
- This will need CI builds and tests. Review the following resources on cuDF's Java CI and try to duplicate them:
  - [GitHub Action Job for PRs](https://github.com/rapidsai/cudf/blob/fa8284ddb2de808573d5b21cc9e650578ddf6acc/.github/workflows/pr.yaml#L82-L91)
  - [GitHub Action Job for Nightly Tests](https://github.com/rapidsai/cudf/blob/fa8284ddb2de808573d5b21cc9e650578ddf6acc/.github/workflows/test.yaml#L74-L85)
  - [Java build/test script](https://github.com/rapidsai/cudf/blob/branch-24.08/ci/test_java.sh)
    - This uses conda to supply the libcudf package, and builds the Java bindings from that. If you use direct bindings and do not depend on the kvikio C++ library, then you can do something simpler that only requires installing cuFile from conda.
  - [Java dependencies.yaml file key](https://github.com/rapidsai/cudf/blob/fa8284ddb2de808573d5b21cc9e650578ddf6acc/dependencies.yaml#L51-L59)
    - This file key gives a list of what dependency lists to include. This file key is [referenced](https://github.com/rapidsai/cudf/blob/fa8284ddb2de808573d5b21cc9e650578ddf6acc/ci/test_java.sh#L12-L15) by the CI build/test script. This file key should include [the dependency list `cuda`](https://github.com/rapidsai/kvikio/blob/157a29a8530494998642a9983aef27f06bd94b67/dependencies.yaml#L191-L241), so you get the cufile package. See the [rapids-dependency-file-generator README](https://github.com/rapidsai/dependency-file-generator) for more information.
  - [Java dependencies.yaml list](https://github.com/rapidsai/cudf/blob/fa8284ddb2de808573d5b21cc9e650578ddf6acc/dependencies.yaml#L641-L647)
    - This dependency list should probably include `maven` and `openjdk`. Not sure what else.

Give this a start and please feel free to schedule time with me to discuss anything that is unclear!

jakirkham commented:
Is this still planned for 24.08 or should we move to 24.10?

aslobodaNV commented:
I will not likely finish the updates here for a couple weeks due to other priorities, so this should be moved to 24.10

jakirkham commented:
Ok thanks Alex! 🙏

Have moved to 24.10

bdice commented on .github/workflows/build.yaml:
```diff
@@ -34,6 +34,14 @@ jobs:
       branch: ${{ inputs.branch }}
       date: ${{ inputs.date }}
       sha: ${{ inputs.sha }}
+  java-build:
```
`build.yaml` is only used for building packages that we distribute as a part of nightly builds. Do you expect this to distribute a new conda package? It should probably be removed, see other comments below.

aslobodaNV replied:
I don't think we expect this to distribute a new conda package, so I have removed this bit

bdice commented on .github/workflows/pr.yaml:
```diff
@@ -39,6 +41,18 @@ jobs:
     uses: rapidsai/shared-workflows/.github/workflows/conda-cpp-tests.yaml@branch-24.10
     with:
       build_type: pull-request
+  conda-java-build:
+    needs: checks
+    secrets: inherit
+    uses: rapidsai/shared-workflows/.github/workflows/conda-java-build.yaml@branch-24.10
+    with:
+      build_type: pull-request
+  conda-java-tests:
+    needs: conda-java-build
+    secrets: inherit
+    uses: rapidsai/shared-workflows/.github/workflows/conda-java-tests.yaml@branch-24.10
+    with:
+      build_type: pull-request
```
There are [no conda-java-build or conda-java-tests workflows](https://github.com/rapidsai/shared-workflows/tree/branch-24.10/.github/workflows). We use a single "custom job" workflow for Java builds/tests in cuDF. Copy from here: https://github.com/rapidsai/cudf/blob/afd3a4b4776adf738284c9f0b99e1fc2fcefeec8/.github/workflows/pr.yaml#L153-L163

aslobodaNV replied:
Moved to use custom job

bdice commented on .github/workflows/test.yaml:
```diff
@@ -30,3 +30,11 @@ jobs:
       branch: ${{ inputs.branch }}
       date: ${{ inputs.date }}
       sha: ${{ inputs.sha }}
+  java-tests:
```
We do run Java tests in nightlies, so this can stay, but it needs to use a custom job.

aslobodaNV replied:
Added custom job

bdice commented on java/README.md:
```diff
@@ -0,0 +1,74 @@
+#### Run example
+
+    cd kvikio/java/
+    java -cp target/cufile-24.10.0-SNAPSHOT.jar:$HOME/.m2/repository/org/jcuda/jcuda/12.0.0/jcuda-12.0.0.jar:$HOME/.m2/repository/org/jcuda/jcuda-natives/12.0.0/jcuda-natives-12.0.0.jar -Djava.library.path=./target bindings.kvikio.example.Main
```
> `cufile-24.10.0-SNAPSHOT.jar`

cuFile is not versioned like RAPIDS. Is this correct?

aslobodaNV replied:
Regarding the version, the jar that is generated is not aware of what version of libcufile was used to generate it... that just depends on what version of the cuda toolkit the host machine has installed. It appears I can't make the jar not have a version number at all, so I'm inclined to have the version represent the version of kvikio. If you think there's a way I can inject the version of libcufile into maven at runtime I am open to that, but I have not found any information on how that would be done so far.

bdice commented on java/src/main/java/bindings/kvikio/cufile/CuFileDriver.java:
```diff
@@ -0,0 +1,35 @@
+ * limitations under the License.
+ */
+
+ package bindings.kvikio.cufile;
```
Is there code formatting that can be applied? There are some extraneous spaces in this PR.
```suggestion
package bindings.kvikio.cufile;
```

aslobodaNV replied:
Thanks, these should be resolved now.

bdice reviewed (commented):
Quick first pass of review -- mostly focused on packaging and CI. Give my suggestions a try and I'll comment `/ok to test` to run your next commits on CI.

aslobodaNV commented:
> Quick first pass of review -- mostly focused on packaging and CI. Give my suggestions a try and I'll comment `/ok to test` to run your next commits on CI.

@bdice I believe I have updated the PR with all of your suggestions, except for the versioning question you had. Please let me know if you had further thoughts there or if it seems that I missed anything!

madsbk commented on java/src/main/native/src/CuFileJni.cpp:
```diff
@@ -0,0 +1,237 @@
+   * @param buffer Device buffer to read the file content into.
+   * @param file_offset Starting offset from which to read the file.
+   */
+  void read(void* buffer,
+            std::size_t size,
+            std::size_t file_offset,
+            std::size_t device_offset) const
+  {
+    auto const status = cuFileRead(cufile_handle_, buffer, size, file_offset, device_offset);
+
+    if (status < 0) {
+      if (IS_CUFILE_ERR(status)) {
+        throw std::logic_error("Failed to read file into buffer: " + cuFileGetErrorString(status));
+      } else {
+        throw std::logic_error("Failed to read file into buffer: " + cuFileGetErrorString(errno));
+      }
+    }
+  }
+
+  void write(void* buffer, std::size_t size, std::size_t file_offset, std::size_t buffer_offset)
+  {
+    auto const status = cuFileWrite(cufile_handle_, buffer, size, file_offset, buffer_offset);
+    if (status < 0) {
+      if (IS_CUFILE_ERR(status)) {
+        throw std::logic_error("Failed to write file from buffer: " + cuFileGetErrorString(status));
+      } else {
+        throw std::logic_error("Failed to write file from buffer: " + cuFileGetErrorString(errno));
+      }
+    }
+  }
```
Consider adding a `sync_default_stream` argument. 
>   Contrary to most of the non-async CUDA API, cuFile does not have the semantic of being ordered
>   with respect to other non-cuFile work in the default stream. By enabling `sync_default_stream`,
>   KvikIO will synchronize the default stream and order the operation with respect to other work
>   in the null stream. When in KvikIO's compatibility mode or when accessing host memory, the
>   operation is always default stream ordered like the rest of the non-async CUDA API. In this
>   case, the value of `sync_default_stream` is ignored.

See https://github.com/rapidsai/kvikio/blob/branch-24.10/cpp/include/kvikio/file_handle.hpp#L317-L323

aslobodaNV replied:
Hey Mads, just saw this. I think this is a good suggestion and one we should definitely incorporate. My bandwidth is very limited at the moment so I am unable to add it now, would you be okay with adding this in a future update? I think this will require some updates to the maven file to support the proper compilation with the cuda libraries and it takes me a good amount of time to context switch+remember how maven works.

bdice commented on .github/workflows/build.yaml:
```diff
@@ -44,7 +44,7 @@ jobs:
       sha: ${{ inputs.sha }}
   upload-conda:
-    needs: [cpp-build, python-build]
+    needs: [cpp-build, python-build, java-build]
```
Revert, since no `java-build` job exists anymore.
```suggestion
    needs: [cpp-build, python-build]
```

bdice commented on .github/workflows/pr.yaml:
```diff
@@ -39,6 +41,18 @@ jobs:
+  conda-java-build:
+    needs: checks
+    secrets: inherit
+    uses: rapidsai/shared-workflows/.github/workflows/conda-java-build.yaml@branch-24.10
```
Since we aren't building a conda package of this, we do not need a build job. The `conda-java-build` can be deleted, and leave only `conda-java-tests` below.

The test job will:
- Generate a conda environment with all build dependencies
- Install libkvikio
- Run maven to build the Java bindings and tests
- Run the tests

bdice commented on .github/workflows/pr.yaml:
```diff
@@ -39,6 +41,18 @@ jobs:
+  conda-java-tests:
+    needs: conda-java-build
+    secrets: inherit
+    uses: rapidsai/shared-workflows/.github/workflows/conda-java-tests.yaml@branch-24.10
```
This is the other place we need to use a custom job. [There is no shared-workflows workflow named `conda-java-tests`](https://github.com/rapidsai/shared-workflows/tree/branch-24.12/.github/workflows). Make this look like the code in `test.yaml` except with different parameters (different `build_type:`, no `branch:`, etc.). Look at cuDF for inspiration here.

Also, let's retarget this work to branch-24.12. You'll need to:
- Merge the latest `branch-24.12` into your PR
- Update any lingering references to `branch-24.10` (this line will be one)
- Change the target branch on your PR to `branch-24.12`

bdice commented on .github/workflows/test.yaml:
```diff
@@ -30,3 +30,15 @@ jobs:
       sha: ${{ inputs.sha }}
+  conda-java-tests:
+    secrets: inherit
+    uses: rapidsai/shared-workflows/.github/workflows/custom-job.yaml@python-3.12
```
Use `branch-24.12` after you merge in the upstream and retarget this PR:
```suggestion
    uses: rapidsai/shared-workflows/.github/workflows/custom-job.yaml@branch-24.12
```

bdice commented on dependencies.yaml:
```diff
@@ -88,6 +88,11 @@ files:
       - test_python
+  test_java:
+    output: none
+    includes:
+      - cuda
+      - test_java
```
I'm guessing you'll need CMake and C++ compilers? You'll also need `cuda_version` to constrain which nvcc and nvcomp versions you get.
```suggestion
    includes:
      - build-universal
      - build-cpp
      - cuda_version
      - cuda
      - test_java
```

bdice reviewed (commented):
Keeping this pass of review briefer and to-the-point: once we get `pr.yaml` and `dependencies.yaml` fixed up, then we can start running CI and see how much is broken.

bdice commented on dependencies.yaml:
```diff
@@ -336,3 +344,15 @@ dependencies:
           - matrix: # All CUDA 11 versions
             packages:
               - cuda-python>=11.7.1,<12.0a0
+  test_python_legate:
+    common:
+      - output_types: [conda, requirements, pyproject]
+        packages:
+          - *dask
+          - distributed>=2022.05.2
```
This is probably from an improper merge conflict resolution. I think we can delete it.
```suggestion
```

bdice reviewed (commented):
Okay, I think we're at the point where I can trigger a CI run and we can learn from whatever goes wrong.

bdice commented:
/ok to test

bdice commented on .github/workflows/pr.yaml:
```diff
@@ -39,6 +40,15 @@ jobs:
     with:
       build_type: pull-request
+  conda-java-tests:
+    secrets: inherit
```
This job should run _after_ the libkvikio conda packages are built, [since it uses those](https://github.com/rapidsai/kvikio/pull/396/files#diff-76d7916ab418397eef76faf7d4eb5f000cff2b417def935c97f0d2c727afb6afR26-R28).
```suggestion
    needs: conda-cpp-build
    secrets: inherit
```

bdice commented on ci/test_java.sh:
This may need a `chmod +x` to be runnable. I'll push a couple changes to this PR.

```
/__w/_temp/90874e9a-0325-462e-b70a-5fe0c2a4e2bc.sh: line 1: ci/test_java.sh: Permission denied
Error: Process completed with exit code 126.
```

bdice commented:
/ok to test

bdice commented:
/ok to test

bdice commented on java/pom.xml:
```diff
@@ -0,0 +1,147 @@
+            <configuration>
+              <target>
+                <!-- Compile native code using nvcc -->
+                <exec executable="/usr/local/cuda/bin/nvcc">
```
@aslobodaNV The first real failure: this path is not the correct path to `nvcc` in the CI environment, which installed `cuda-nvcc` with `conda`. Can you just call `nvcc`?
```suggestion
                <exec executable="nvcc">
```

[Error log here:](https://github.com/rapidsai/kvikio/actions/runs/11134675532/job/30943482558?pr=396#step:9:688)
```
Error:  Failed to execute goal org.apache.maven.plugins:maven-antrun-plugin:3.0.0:run (compile-native-code) on project cufile: An Ant BuildException has occured: Execute failed: java.io.IOException: Cannot run program "/usr/local/cuda/bin/nvcc" (in directory "/__w/kvikio/kvikio/java"): error=2, No such file or directory
Error:  around Ant part ...<exec executable="/usr/local/cuda/bin/nvcc">... @ 4:49 in /__w/kvikio/kvikio/java/target/antrun/build-main.xml
```

bdice replied:
Similarly, the include paths should be located in the conda environment. The CUDA packages use a special "targets" layout. For example, `libcufile-dev` will install `cufile.h` into `${CONDA_PREFIX}/targets/x86_64-linux/include/cufile.h`.

The `openjdk` package installs some headers into `${CONDA_PREFIX}/include` and some into `${CONDA_PREFIX}/include/linux`.

CMake knows how to detect and parse this layout of the CUDA Toolkit but I do not know how Maven or other build systems are supposed to consume it. I would ask the Spark team, perhaps they know: @revans2, do you have insight here?

aslobodaNV replied:
Totally fine with committing the update to just call nvcc instead of the path, I just didn't happen to have nvcc on my path when I wrote that. Let me look into the include paths and reach out to Bobby if I get stuck.

bdice commented on dependencies.yaml:
```diff
@@ -353,3 +361,9 @@ dependencies:
+      - output_types: conda
+        packages:
+          - maven
+          - openjdk=8.*
```
Is there a reason to use `openjdk=8.*`? It appears that the latest is `openjdk=22.*`.

aslobodaNV replied:
No particular reason, I just happened to have 8 installed and used that. I will update to 22

revans2 replied:
JDK8 is a very much a least common denominator. It is no longer supported without paying a lot of money to oracle though.

https://www.oracle.com/java/technologies/java-se-support-roadmap.html

and oracle is trying very hard to push people away from older versions of java. 

java 11, 17, and 21 are long term support versions that are currently GA.

Java generally has really good backwards compatibility so almost anything compiled with JDK8 can run on anything newer. But newer code is not backwards compatible with older runtimes. But you can ask newer compilers to output binary code that is compatible with older JDK versions. That is what we do with the Spark plugin.

We also support whatever Spark supports for their default binary release. So for older versions of Spark it is JDK8. For some of the latest versions that is JDK17, but because of how we ask the compiler to output older binary versions it really becomes a requirement that they have at least that version of java installed.

I wouldn't update to jdk22 unless you also decide which version of java you want to be minimally compatible with and ask the compiler to output compatible .class files.

Also just another point to think about jcuda still uses openjdk8.  This is probably to have as much compatibility as possible. https://github.com/jcuda/jcuda-main/blob/master/BUILDING.md

aslobodaNV replied:
Thanks for that context, that makes a lot of sense. I have not been following the broader Java versioning ecosystem so was unaware of much of this. I think it is likely we will want to pick one of the versions with long term support, I would guess 17 or 21, but will think on it some more before making an update.

revans2 commented on java/src/main/java/bindings/kvikio/cufile/CuFile.java:
```diff
@@ -0,0 +1,45 @@
+ * limitations under the License.
+ */
+
+package bindings.kvikio.cufile;
```
nit: The convention for java packages is to include the organization/web address in reverse to avoid conflicts with others. Is there reason this isn't `ai.rapids.kvikio.cufile` similar to how the cudf bindings are `ai.rapids.cudf`?

aslobodaNV replied:
No particular reason, let me update that

revans2 commented on java/src/main/java/bindings/kvikio/cufile/CuFileHandle.java:
```diff
@@ -0,0 +1,39 @@
+
+package bindings.kvikio.cufile;
+
+abstract class CuFileHandle implements AutoCloseable {
```
It might be nice to have some javadocs here.

aslobodaNV replied:
I will be sure to get these javadocs that you have requested in before we merge, just wanted to tackle the build issues we were seeing first. Thank you for all your feedback.

revans2 commented on java/src/main/java/bindings/kvikio/cufile/CuFileReadHandle.java:
```diff
@@ -0,0 +1,34 @@
+
+package bindings.kvikio.cufile;
+
+public final class CuFileReadHandle extends CuFileHandle {
```
javadocs would be very nice for this class as it is intended to be used by developers.

revans2 commented on java/src/main/java/bindings/kvikio/cufile/CuFileWriteHandle.java:
```diff
@@ -0,0 +1,33 @@
+
+package bindings.kvikio.cufile;
+
+public final class CuFileWriteHandle extends CuFileHandle {
```
Same here javadocs would be very nice for the developer experience.

revans2 reviewed (commented):
Overall it looks good. Just a few nits. Mostly around java convention. I am curious about your plans to distribute binaries. The java code itself is going to require the native library to work, but the native library is something that the user is expected to compile themselves. This is not a great user experience, but it is a real pain to try and distribute native binaries too. If you want/need some help with this there are some people on my team that have spent a lot of time to make this work well, especially for x86 linux.

aslobodaNV commented:
@madsbk or @bdice would one of you be able to try and run this through the build tests? I have updated it to try and make it more agnostic to the build location. It works on my machine, but need to know if there are more build errors remaining to be fixed for the CI/CD machines.

revans2 commented on java/pom.xml:
```diff
@@ -0,0 +1,150 @@
+          <version>3.8.0</version>
+          <configuration>
+                <source>21</source>
+                <target>21</target>
```
This is where you can set what the output target is that you want.

revans2 reviewed (commented):
Generally it looks really good. I am not going to approve it because I don't know enough about cufile to be able to say that this matches all of the requirements. But from a java perspective it looks decent.

aslobodaNV commented on java/src/main/java/bindings/kvikio/cufile/CuFile.java:
```diff
@@ -0,0 +1,45 @@
+                    driver.close();
+                }));
+                initialized = true;
+            } catch (Throwable t) {
```
Would you be able to point me to an example of best practices in this area? Happy to change it further from my latest update.

revans2 replied:
I marked it as a nit because it is probably fine in this case. In fact when I went back to look at the CUDF code, I did the exact same thing 5 years ago.

https://github.com/rapidsai/cudf/blob/4dbb8a354a9d4f0b4d82a5bf9747409c6304358f/java/src/main/java/ai/rapids/cudf/NativeDepsLoader.java#L74-L83

It is just generally not good practice to catch a `Throwable`, because `Throwable` includes all `Error`s and `Error`s are generally considered to not be recoverable. But really the main thing here is giving the user enough information that they can debug why it is failing. You are printing the error message, but the full stack trace would be better. That is the main thing that I would suggest you do.

bdice commented on java/src/main/native/CMakeLists.txt:
```diff
@@ -0,0 +1,21 @@
+cmake_minimum_required(VERSION 3.10)
```
RAPIDS requires a minimum of cmake 3.26.4 right now, I think. Can we set a higher minimum for these bindings? I don't think we will ever test with a CMake version as old as 3.10.

bdice commented:
/ok to test

bdice commented:
/ok to test

bdice commented:
/ok to test

revans2 reviewed (approved):
From the java side of things it looks good. I am not an expert on the APIs or all of the use cases, but the plumbing looks decent.

bdice commented:
@aslobodaNV Looks like there are CI build failures. I'm not sure how to diagnose this.

```
Error:  Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0:compile (default-compile) on project cufile: Fatal error compiling: error: invalid target release: 21 -> [Help 1]
```

bdice commented on java/pom.xml:
```diff
@@ -0,0 +1,150 @@
+
+  <properties>
+    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    <maven.compiler.source>21</maven.compiler.source>
+    <maven.compiler.target>21</maven.compiler.target>
```
Maybe the build failures are expecting Java 21, but only 11 is present?
```suggestion
    <maven.compiler.source>11</maven.compiler.source>
    <maven.compiler.target>11</maven.compiler.target>
```

aslobodaNV replied:
Yes this problem I was able to fix, there is another linking issue with the libraries that I am running into now though. Let me push up my latest updates

bdice commented:
/ok to test

bdice commented:
/ok to test

madsbk commented on java/src/main/native/src/CuFileJni.cpp:
```diff
@@ -0,0 +1,238 @@
+  cufile_file& operator=(cufile_file const&) = delete;
+
+  /** @brief Destroy the file wrapper by de-registering the cuFile handle and closing the file. */
+  ~cufile_file()
```
```suggestion
  ~cufile_file() noexcept
```

madsbk commented on java/src/main/native/src/CuFileJni.cpp:
```diff
@@ -0,0 +1,238 @@
+  ~cufile_file()
+  {
+    cuFileHandleDeregister(cufile_handle_);
+    close(file_descriptor_);
```
Maybe print a warning if `close` fails?

aslobodaNV replied:
I would prefer to avoid printing to std::cerr from here. I think it would be good in a future PR to add some improved logging in general though and this could be one spot for that

madsbk commented on java/src/main/java/ai/rapids/kvikio/cufile/CuFileReadHandle.java:
```diff
@@ -0,0 +1,59 @@
+ * offset, transferring it to a device buffer.
+ * </p>
+ */
+public final class CuFileReadHandle extends CuFileHandle {
```
Consider merging `CuFileReadHandle` and `CuFileWriteHandle` into a single class. Since it is possible to open a file in read+write mode, I am not sure having two distinct classes is that useful. E.g., we would need a `CuFileWriteHandle.read()` and a `CuFileReadHandle.write()`?

aslobodaNV replied:
It is possible, however in Java it appears the canonical way to manage files is via separate read+write handles, as opposed to a shared reader+writer. There does exist this `RandomAccessFile` class in Java that supports reads+writes: https://docs.oracle.com/en/java/javase/18/docs/api/java.base/java/io/RandomAccessFile.html

My personal opinion is it would be best to support 3 versions, Read, Write, and an equivalent to `RandomAccessFile` since that appears to be more in line with Java best practices.

madsbk replied:
Fair enough :)

madsbk commented:
/ok to test

madsbk reviewed (approved):
Looks good, thanks @aslobodaNV

bdice commented on ci/test_java.sh:
```diff
@@ -0,0 +1,47 @@
+
+rapids-mamba-retry install \
+  --channel "${CPP_CHANNEL}" \
+  libkvikio libkvikio-tests
```
We probably don’t need the tests package.
```suggestion
  libkvikio
```

bdice commented on dependencies.yaml:
```diff
@@ -388,3 +396,12 @@ dependencies:
+          - *cmake_ver
+          - maven
+          - openjdk=11.*
+          - make
```
Alphabetize this, put make above maven.

bdice commented on dependencies.yaml:
```diff
@@ -388,3 +396,12 @@ dependencies:
+    common:
+      - output_types: conda
+        packages:
+          - cxx-compiler
```
cxx-compiler and cmake can be removed. We are already including the build-universal and build-cpp lists in the test_java file key, so they will be included that way.

bdice commented on java/README.md:
```diff
@@ -0,0 +1,74 @@
+# Java KvikIO Bindings
+
+## Summary
+These Java KvikIO bindings for GDS currently support only synchronous read and write IO operations using the underlying CuFile API. Support for batch IO and asynchronous operations are not yet supported.
```
For consistency, use a lowercase c in cuFile. That matches the official docs. Please apply this everywhere in your PR.
```suggestion
These Java KvikIO bindings for GDS currently support only synchronous read and write IO operations using the underlying cuFile API. Support for batch IO and asynchronous operations are not yet supported.
```

bdice commented on java/README.md:
```diff
@@ -0,0 +1,74 @@
+    -Djava.library.path={path/to/your/so/file/}
+
+## Examples
+An example for how to use the Java KvikIO bindings can be found in src/main/java/bindings/kvikio/example . Note: This example has a dependency on JCuda so ensure that when running the example the JCuda shared library files are on the JVM library path along with the libCuFileJNI.so file.
```
Let’s sprinkle some backticks around all code-like text in this document so it renders as code font.
```suggestion
An example for how to use the Java KvikIO bindings can be found in `src/main/java/bindings/kvikio/example`. Note: This example has a dependency on JCuda so ensure that when running the example the JCuda shared library files are on the JVM library path along with the libCuFileJNI.so file.
```

bdice commented on java/README.md:
```diff
@@ -0,0 +1,74 @@
+    cd kvikio/java/src/main/java/bindings/kvikio/cufile
+    javac *.java
+
+#### Retrieve Jcuda jar files
```
```suggestion
#### Retrieve JCuda jar files
```

bdice commented on java/pom.xml:
```diff
@@ -0,0 +1,146 @@
+  <artifactId>cufile</artifactId>
+  <version>24.12.0-SNAPSHOT</version>
+
+  <name>cufile</name>
```
```suggestion
  <name>cuFile</name>
```

bdice commented on java/pom.xml:
```diff
@@ -0,0 +1,146 @@
+      This project provides Java bindings for the GPUDirect Storage cufile library, enabling the GPU to load and
+      save large amounts of data to and from persistent storage. This is still a work in progress so some APIs may change.
+  </description>
+  <url>http://ai.rapids</url>
```
```suggestion
  <url>https://rapids.ai</url>
```

aslobodaNV replied:
Is this correct? I copied from here:
https://github.com/rapidsai/cudf/blob/4da6fda3e6042645b8e21c931b26966ef0fa8897/java/pom.xml#L31

bdice replied:
I think this is an error in cudf. ai.rapids is not the homepage URL. It’s rapids.ai.

bdice commented on java/src/main/native/src/CuFileJni.cpp:
```diff
@@ -0,0 +1,238 @@
+#include <unistd.h>
+#include <memory>
+
+#include <stdexcept>
```
Why are there blank lines between standard library headers? Typically we group those together and alphabetize them. Maybe check that clang-format is being applied. It may be the case that we don’t have include groupings in the clang-format settings for this repo.

aslobodaNV replied:
My local formatter must have not worked on this file. I have updated it

bdice commented on java/src/main/native/src/CuFileJni.cpp:
```diff
@@ -0,0 +1,238 @@
+   * @brief Factory method to create a file wrapper for writing.
+   *
+   * @param path Absolute path of the file to write to. This creates the file if it does not already
+   * exist..
```
```suggestion
   * exist.
```

bdice commented:
@aslobodaNV Great work. I left some comments but we should be able to get this merged after they are addressed.

madsbk commented:
> @madsbk Do you want to target 24.12 or 25.02?

I have no strong opinion but it might make sense to add some more features before we release? 
@aslobodaNV do you want this to part of the 24.12 release? Or should we have it in the nightly channel until 25.02?

aslobodaNV commented:
> @aslobodaNV Great work. I left some comments but we should be able to get this merged after they are addressed.

@bdice Addressed all comments except the rapids url change. Please confirm that it should differ from the one used in the cudf repo.

bdice reviewed (approved):
Approving and retargeting to 25.02.

bdice commented:
One last CI run, then this should be good to go.

bdice commented:
/ok to test

bdice commented:
/merge

bdice commented:
/ok to test

bdice commented:
/ok to test

bdice commented:
/merge
